x-pack/plugins/apm/public/components/alerting/ui_components/chart_preview/chart_preview_helper.tsx

```diff
@@ -0,0 +1,159 @@
+      return niceTimeFormatter([xMin, xMax]);
+    } else {
+      return (value: number) => `${value}`;
+    }
+  }, [xMin, xMax]);
+  return dateFormatter;
+};
+
+export const getDomain = (
+  series: Array<{ name?: string; data: Coordinate[] }>
+) => {
+  let min: number | null = null;
+  let max: number | null = null;
+  const valuesByTimestamp = series.reduce<{ [timestamp: number]: number[] }>(
+    (acc, serie) => {
+      serie.data.forEach((point) => {
+        const valuesForTimestamp = acc[point.x] || [];
+        acc[point.x] = [...valuesForTimestamp, point.y ?? 0];
+      });
+      return acc;
+    },
+    {}
+  );
+  const pointValues = Object.values(valuesByTimestamp);
+  pointValues.forEach((results) => {
+    const maxResult = getMax(results);
+    const minResult = getMin(results);
+    if (maxResult && (!max || maxResult > max)) {
+      max = maxResult;
+    }
+    if (minResult && (!min || minResult < min)) {
+      min = minResult;
+    }
+  });
+  const timestampValues = Object.keys(valuesByTimestamp).map(Number);
+  const minTimestamp = getMin(timestampValues) || 0;
+  const maxTimestamp = getMax(timestampValues) || 0;
+  return {
+    yMin: min || 0,
+    yMax: max || 0,
+    xMin: minTimestamp,
+    xMax: maxTimestamp,
+  };
+};
+
+const EmptyContainer: React.FC = ({ children }) => (
+  <div
+    style={{
+      width: '100%',
+      height: 150,
+      display: 'flex',
+      justifyContent: 'center',
+      alignItems: 'center',
+    }}
+  >
+    {children}
+  </div>
+);
+
+export function NoDataState() {
+  return (
+    <EmptyContainer>
+      <EuiText color="subdued" data-test-subj="noChartData">
+        <FormattedMessage
+          id="xpack.apm.alerts.charts.noDataMessage"
+          defaultMessage="No chart data available"
+        />
+      </EuiText>
+    </EmptyContainer>
+  );
+}
+
+export function LoadingState() {
+  return (
+    <EmptyContainer>
+      <EuiText color="subdued" data-test-subj="loadingData">
+        <EuiLoadingChart size="m" />
+      </EuiText>
+    </EmptyContainer>
+  );
+}
+
+export function ErrorState() {
+  return (
+    <EmptyContainer>
+      <EuiText color="subdued" data-test-subj="chartErrorState">
+        <FormattedMessage
+          id="xpack.apm.alerts.charts.errorMessage"
+          defaultMessage="Uh oh, something went wrong"
+        />
+      </EuiText>
+    </EmptyContainer>
+  );
+}
+
+interface PreviewChartLabel {
+  lookback: number;
+  timeLabel: string;
+  displayedGroups: number;
+  totalGroups: number;
+}
+
+export function TimeLabelForData({
+  lookback,
+  timeLabel,
+  displayedGroups,
+  totalGroups,
+}: PreviewChartLabel) {
+  return (
+    <div style={{ textAlign: 'center' }}>
+      <EuiText size="xs" color="subdued">
+        <FormattedMessage
+          id="xpack.apm.alerts.timeLabelForData"
+          defaultMessage="Last {lookback} {timeLabel} of data, showing {displayedGroups}/{totalGroups} groups"
+          values={{
+            lookback,
```

Comment:
I don't think `lookback` is needed here. It can be derived from `series` afaict

Reply:
You can calculate the lookback like this: 

```ts
const previewDuration = moment.duration(xMax - xMin).asMinutes()l
```

Reply:
Interestingly, calculating from xMax and `xMin` will also give more correct results. The chart below shows data from 21:15 to 21:30 = 15min. Yet the label shows "25 min":

![image](https://github.com/elastic/kibana/assets/209966/0f7069c1-f7e9-48ce-b817-e19df8363cc0)

Reply:
The total data should still be 25 minutes. If you hove over last bar in last timestamp, that data should be after 21:30:00.

Reply:
in this case, were there data for last 25 minutes present?

Reply:
> The total data should still be 25 minutes. If you hove over last bar in last timestamp, that data should be after 21:30:00.

Not sure what you mean. This says 5 minutes but shows data between 23:08 to 23:10.

![Kapture 2023-06-14 at 23 13 25](https://github.com/elastic/kibana/assets/209966/ffa4933e-7e7c-4355-8c1f-6d5045d102e0)

Reply:
> in this case, were there data for last 25 minutes present?

Yes, there was data. Either way, I don't think it matters since we use [`extended_bounds`](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-histogram-aggregation.html#search-aggregations-bucket-histogram-aggregation-extended-bounds)

Reply:
on CCS, it looks fine. :thinking:

<img width="593" alt="Screenshot 2023-06-14 at 23 26 11" src="https://github.com/elastic/kibana/assets/69037875/bb4fd54a-f232-4812-844e-cca4f1b90d65">

Reply:
> in this case, were there data for last 25 minutes present?

I'm not sure it matters since we use [`extended_bounds`](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-histogram-aggregation.html#search-aggregations-bucket-histogram-aggregation-extended-bounds).

We should also show empty buckets.

Reply:
Well.. either way, no need to rely on `BUCKET_SIZE` to calculate the duration/lookback.

Reply:
>  The chart below shows data from 21:15 to 21:30 = 15min. Yet the label shows "25 min":

could you tell me which dataset you were using?

Reply:
I was using the `simple_trace` scenario

Reply:
The issue in this case is that if there is no data in the beginning or end of period, the x-axis doesn't show that timestamp. I think in this case we should show the timestamps even if there is no data for that time instead of calculating time based on what is displayed. It would be good to show that there is no data in that time.

I tested with `simple_trace` scenario which I ran multiple times to generate data points at different intervals.
<img width="604" alt="Screenshot 2023-06-19 at 12 02 48" src="https://github.com/elastic/kibana/assets/69037875/d0a26b86-7d98-4d41-94a5-54493437821b">

I will create a follow up issue and work on it in a separate PR, unless you disagree.

Reply:
> I will create a follow up issue and work on it in a separate PR, unless you disagree.

That sounds good to me 👍

Reply:
created https://github.com/elastic/kibana/issues/159928